src/hotspot/share/utilities/ostream.hpp

```diff
@@ -56,6 +58,20 @@ class outputStream : public CHeapObjBase {
 
   // Returns whether a newline was seen or not
    bool update_position(const char* s, size_t len);
+
+  // Nominally processes the given format string and the supplied arguments
+  // to produce a formatted string using the supplied buffer. The formatted
+  // string (in the buffer) is returned, and the outgoing `result_len` set
+  // to the size of the returned string.
+  //
+  // If the format string is a plain string (no format specifiers)
+  // or idiomatically it is "%s" to print a supplied argument string, then
+  // the buffer is ignored, we return the string directly and set `result_len` to its
+  // length. However, if `add_cr` is true then we have to copy the string
+  // into the buffer and we risk truncation. The `result_len` is always set to the length
+  // of the returned string.
+  //
+  // In a debug build, if truncation occurs a VM warning is issued.
```

Comment:
I had to think a bit (I am not a native English speaker) about what the "Nominally" means, but I think it is supposed to contrast the second paragraph? As in "Normally we do that, but in the case of ... we do... ?". Same for "idiomatically" - what does that signify?

Reply:
Right "nominally" is indicating that it basically operates one way but there are exceptions as outlined in the second paragraph.

"idiomatically" means we are applying a specific coding idiom aka pattern - in this case secure programming says you never pass a non-constant string to a printf-like function, but instead pass "%s" and supply the actual string as the argument. So when we encounter that idiom will handle it specially.

Reply:
> result_len contains the length of the returned string, including trailing cr, excluding trailing nul

Yes I state that it is the length of the returned string as per strlen.

Yes I can clarify cr is always printed.

> It may bypass the caller-provided output buffer, returning the provided format string directly, if no formatting or copying is needed

Well it _will_ bypass in those cases, though you then need to know when "no copying is needed". The idea was to clearly document the behaviour so that people would no longer be surprised or make assumptions/interepretations.